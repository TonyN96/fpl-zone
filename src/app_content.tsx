import React from "react";
import { useQuery } from "react-query";
import { BrowserRouter as Router,Route, Routes } from "react-router-dom";
import { useMediaQuery } from "@mui/material";
import { getAllFixtures, getGameData } from "api/fpl_api_provider";
import {
  AccountPage,
  AnalysisPage,
  FixturesAndResultsPage,
  GameweekLivePage,
  LoginPage,
  MyTeamPage,
  RegisterPage,
  ResetPage
} from "pages";
import PrivateRoute from "private_route";
import { AppData } from "types";

import { Logout } from "components/authentication";
import { Notifier, Startup } from "components/layout";

export const AppDataContext = React.createContext<AppData | null>(null);

export default function AppContent(): JSX.Element {
  // Fetching game data which will be made available throughout the app via context provider
  const {
    data: gameData,
    isError: gameDataIsError,
    error: gameDataError,
    isLoading: gameDataIsLoading
  } = useQuery("game-data", getGameData);

  // Fetching fixture data which will be made available throughout the app via context provider
  const {
    data: fixtureData,
    isError: fixtureDataIsError,
    error: fixtureDataError,
    isLoading: fixtureDataIsLoading
  } = useQuery("all-fixtures", getAllFixtures);

  const isCompact = useMediaQuery("(max-width:1500px)");

  const isLoading = gameDataIsLoading || fixtureDataIsLoading;
  const isError = gameDataIsError || fixtureDataIsError;
  const appData = gameData && fixtureData ? { ...gameData, fixtures: fixtureData, isCompact } : null;

  if (isLoading) {
    // Display loading message if data is still being fetched
    return (
      <Startup>
        <Notifier />
      </Startup>
    );
  } else if (isError) {
    // Display error message if data fetch failed
    const error = gameDataError || fixtureDataError;
    const errorMessage = error instanceof Error ? `: ${error.message}` : ".";

    return (
      <Startup>
        <Notifier type='error' message={`An error has occured: ${errorMessage}`} />
      </Startup>
    );
  }

  // Otherwise, render the app
  return (
    <AppDataContext.Provider value={appData}>
      <Router>
        <Routes>
          <Route path='*' element={<LoginPage />} />
          <Route path='/' element={<LoginPage />} />
          <Route path='/login' element={<LoginPage />} />
          <Route path='/register' element={<RegisterPage />} />
          <Route path='/reset' element={<ResetPage />} />
          <Route path='/logout' element={<Logout />} />
          <Route path='/account' element={<PrivateRoute component={<AccountPage />} />} />
          <Route
            path='/gameweek-live'
            element={<PrivateRoute component={<GameweekLivePage />} />}
          />
          <Route path='/my-team' element={<PrivateRoute component={<MyTeamPage />} />} />
          <Route
            path='/fixtures-and-results'
            element={<PrivateRoute component={<FixturesAndResultsPage />} />}
          />
          <Route path='/analysis' element={<PrivateRoute component={<AnalysisPage />} />} />
        </Routes>
      </Router>
    </AppDataContext.Provider>
  );

}
