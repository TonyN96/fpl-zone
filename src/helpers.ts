import { Gameweek, Position, Team } from "types";
import { Player } from "types/player";

export const GAME_STATUS_VALUES = {
  GAME_UPDATING: "Game is updating..",
  SEASON_FINISHED: "The current FPL season has finished, check back next season!",
  GAME_OK: "GAME_OK"
};

// Getting a player by their id
export const GetPlayerById = (playerId: number, players: Player[]): Player => {
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return players.find((player) => player.id === playerId)!;
};

// Getting a team by their id
export const getTeamById = (teamId: number, teams: Team[]): Team => {
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return teams.find((team) => team.id === teamId)!;
};

// Getting a player by their id
export const getPositionById = (positionId: number, positions: Position[]): Position => {
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return positions.find((pos) => pos.id === positionId)!;
};

export const getPlayerImageUrl = (player?: Player): string => {
  if (!player) {
    return `${process.env.PUBLIC_URL}/assets/images/player-placeholder.png`;
  }

  const imgId = player.photo.replace(".jpg", "");

  return `https://resources.premierleague.com/premierleague/photos/players/110x140/p${imgId}.png`;
};

export const getNormalizedString = (input: string): string => {
  return input
    .normalize("NFD")
    .replace(/[\u0300-\u036F]/g, "")
    .toUpperCase();
};

// Adding commas to large numbers
// https://stackoverflow.com/questions/2901102/how-to-print-a-number-with-commas-as-thousands-separators-in-javascript
export const numberWithCommas = (x: number): string => {
  return x.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",");
};

// Formatting date to desired format
export const formatDate = (kickOffDateTime: Date): string => {
  return (
    `${kickOffDateTime.toLocaleDateString(navigator.language, { day: "numeric", month: "short" })
    } ${
      kickOffDateTime.toLocaleTimeString(navigator.language, { hour: "2-digit", minute: "2-digit" })}`
  );
};

/*
  The following function is used to check two things:

  1. If the the most recent gameweek deadline was was within the last 1.5 hours
    (If so, FPL data is innacurate as the game is being updated for approx 1.5 hours after
    each gameweek deadline)

  2. If there is no current gameweek then the season is finished - this must be
    relayed back to the user so they understand why data is unavailable
*/
export const checkGameStatus = (gameweeks: Gameweek[]): string => {
  const currentDateTime = new Date();

  const currentGameweek = gameweeks.find(gameweek => gameweek.is_current);
  const nextGameweek = gameweeks.find(gameweek => gameweek.is_next);

  if (!currentGameweek) {
    return GAME_STATUS_VALUES.SEASON_FINISHED;
  }

  /*
    Manually checking if next gamewek deadline has passed as there
    can sometimes be a delay in update of gameweek 'is_next' status
  */
  const relevantGameweek = !nextGameweek || new Date(nextGameweek.deadline_time) > currentDateTime
    ? currentGameweek
    : nextGameweek;

  const gameweekDeadline = new Date(relevantGameweek.deadline_time);
  const timeDifference = currentDateTime.getTime() - gameweekDeadline.getTime();

  // If current time is 1.5 hours (5400000 ms) since deadline, game will be updating
  if (timeDifference < 5400000) {
    return GAME_STATUS_VALUES.GAME_UPDATING;
  }

  return GAME_STATUS_VALUES.GAME_OK;
};

// Adding delay before next function call
export const delay = async (ms: number | undefined): Promise<void> => {
  await new Promise((res) => setTimeout(res, ms));
};
