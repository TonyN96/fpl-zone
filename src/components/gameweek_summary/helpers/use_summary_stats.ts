import { useContext } from "react";
import { AppDataContext } from "app_content";
import { GetPlayerById, numberWithCommas } from "helpers";
import { AppData } from "types";

import { SummaryStatType } from "../gameweek_summary";

export const useSummaryStats = (): SummaryStatType[] => {
  const { gameweeks, players } = useContext(AppDataContext) as AppData;

  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const currentGameweek = gameweeks.find((gw) => gw.is_current)!;

  const starPlayerId = currentGameweek.top_element_info?.id;
  const starPlayer = starPlayerId ? GetPlayerById(starPlayerId, players) : null;

  const mostTransferredIn = currentGameweek.most_transferred_in
    ? GetPlayerById(currentGameweek.most_transferred_in, players)
    : null;

  const mostCaptained = currentGameweek.most_captained
    ? GetPlayerById(currentGameweek?.most_captained, players)
    : null;

  const mostViceCaptained = currentGameweek.most_vice_captained
    ? GetPlayerById(currentGameweek.most_vice_captained, players)
    : null;

  const mostTransferredInCount = mostTransferredIn
    ? numberWithCommas(mostTransferredIn?.transfers_in_event)
    : null;

  const starPlayerScore = currentGameweek.top_element_info
    ? `${currentGameweek.top_element_info?.points} pts`
    : "N/A";

  return ([
    {
      label: "highest score",
      value: `${currentGameweek.highest_score} pts`
    },
    {
      label: "average score",
      value: `${currentGameweek.average_entry_score} pts`
    },
    {
      label: "star player",
      teamCode: starPlayer?.team_code,
      playerName: starPlayer?.web_name || "",
      value: starPlayerScore
    },
    {
      label: "most captained",
      teamCode: mostCaptained?.team_code,
      playerName: mostCaptained ? mostCaptained.web_name : "N/A"
    },
    {
      label: "most vice-captained",
      teamCode: mostViceCaptained?.team_code,
      playerName: mostViceCaptained ? mostViceCaptained.web_name : "N/A"
    },
    {
      label: "most transferred in",
      teamCode: mostTransferredIn?.team_code,
      playerName: mostTransferredIn?.web_name || "",
      value: mostTransferredInCount || ""
    }
  ]);
};
