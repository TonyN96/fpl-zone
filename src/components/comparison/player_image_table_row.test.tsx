import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { AppDataContext } from "app_content";
import { getPlayerImageUrl } from "helpers";
import { mockAppData, mockPlayers } from "test/test_data";

import "@testing-library/jest-dom/extend-expect";

import { PlayerImageTableRow } from ".";

describe("Player image table row tests", () => {
  let players = mockPlayers.slice(0, 3);

  const mockOnAddPlayerClick = jest.fn();

  const createComponent = (): JSX.Element => {
    return (
      <AppDataContext.Provider value={mockAppData}>
        <PlayerImageTableRow
          onAddPlayerClick={mockOnAddPlayerClick}
          onRemovePlayerClick={jest.fn()}
          players={players}
        />
      </AppDataContext.Provider>
    );
  };

  it("Snapshot test", () => {
    const { asFragment } = render(createComponent());
    expect(asFragment()).toMatchSnapshot();
  });

  describe("Player image container", () => {
    describe("if no player prop is present", () => {
      it("has expected style properties", () => {
        render(createComponent());

        const container = screen.getByTestId(`player-image-container-placeholder`);
        const placeHolderImageUrl = `${process.env.PUBLIC_URL}/assets/images/player-placeholder.png`;

        expect(container).toHaveStyle(`background-image: url(${placeHolderImageUrl})`);
        expect(container).toHaveStyle(`cursor: pointer`);
        expect(container).toHaveStyle(`borderRadius: 50%`);
      });

      it("has onClick attribute if no player prop was passed", () => {
        render(createComponent());

        const container = screen.getByTestId("player-image-container-placeholder");
        fireEvent.click(container);

        expect(mockOnAddPlayerClick).toHaveBeenCalledTimes(1);
      });
    });

    describe("if a player prop is present", () => {
      it("has expected style properties", () => {
        render(createComponent());

        players.forEach((player) => {
          const container = screen.getByTestId(`player-image-container-${player.id}`);
          const imageUrl = getPlayerImageUrl(player);

          expect(container).toHaveStyle(`background-image: url(${imageUrl})`);
          expect(container).toHaveStyle(`cursor: auto`);
          expect(container).toHaveStyle(`borderRadius: auto`);
        });
      });

      it("has no onClick attribute if a player prop was passed", () => {
        render(createComponent());

        players.forEach((player) => {
          const container = screen.getByTestId(`player-image-container-${player.id}`);
          fireEvent.click(container);

          expect(mockOnAddPlayerClick).toHaveBeenCalledTimes(0);
        });
      });
    });
  });

  describe("Placeholder image table cell", () => {
    it("displayed when less than 5 five players are selected", () => {
      render(createComponent());

      expect(screen.getByTestId("player-image-container-placeholder")).toBeInTheDocument();
    });

    it("not displayed when 5 five players are selected", () => {
      players = mockPlayers.slice(0, 5);

      render(createComponent());

      expect(screen.queryByTestId("player-image-container-placeholder")).toBeNull();
    });
  });
});
