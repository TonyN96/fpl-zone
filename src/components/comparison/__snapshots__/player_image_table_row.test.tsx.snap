// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player image table row tests Snapshot test 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1bxscn6"
    data-testid="player-image-container-828"
  >
    <div
      class="MuiBox-root css-s16q1a"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-akrvbn-MuiButtonBase-root-MuiIconButton-root"
        data-testid="remove-button"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorInfo MuiSvgIcon-fontSizeMedium css-1yq1soz-MuiSvgIcon-root"
          data-testid="CloseIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    class="MuiBox-root css-10izg3s"
    data-testid="player-image-container-662"
  >
    <div
      class="MuiBox-root css-s16q1a"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-akrvbn-MuiButtonBase-root-MuiIconButton-root"
        data-testid="remove-button"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorInfo MuiSvgIcon-fontSizeMedium css-1yq1soz-MuiSvgIcon-root"
          data-testid="CloseIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    class="MuiBox-root css-1kuydbx"
    data-testid="player-image-container-756"
  >
    <div
      class="MuiBox-root css-s16q1a"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-akrvbn-MuiButtonBase-root-MuiIconButton-root"
        data-testid="remove-button"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorInfo MuiSvgIcon-fontSizeMedium css-1yq1soz-MuiSvgIcon-root"
          data-testid="CloseIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    class="MuiBox-root css-cscpxd"
    data-testid="player-image-container-placeholder"
  >
    <div
      class="MuiBox-root css-1v6v5j6"
      data-testid="add-button"
    >
      <button
        aria-label="add player button"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="AddIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
